Stylecheck script fails on MacOS #13492

hiroshitashir opened the pull request:
- Install GNU grep on macOS with `brew install grep`
- Use GNU grep with -E option
- The following errors fixed.
```
$scripts/check_style.sh
ggrep: warning: stray \ before /
ggrep: warning: stray \ before /
ggrep: warning: * at start of expression
ggrep: warning: * at start of expression
```

The change was tested on macOS and Linux

Fixes https://github.com/ethereum/solidity/issues/13492

hiroshitash commented:
I believe [these](https://app.circleci.com/pipelines/github/ethereum/solidity/30308/workflows/f88c02a0-8229-4ef1-af9d-ffaea84fc778/jobs/1345063/parallel-runs/0/steps/0-103)  are valid style errors. '*'s should be left-aligned. 

```
Coding style error:
libsolidity/analysis/ControlFlowBuilder.cpp:587:	if (auto const *builtinFunction = m_inlineAssembly->dialect().builtin(_functionCall.functionName.name))
libsolidity/analysis/ControlFlowBuilder.h:155:	void placeAndConnectLabel(CFGNode *_node);
libsolidity/analysis/TypeChecker.cpp:1556:			if (auto const *variableDeclaration = dynamic_cast<VariableDeclaration const*>(identifier->annotation().referencedDeclaration))
libsolidity/codegen/CompilerContext.h:376:	CompilerContext *m_runtimeContext;
libsolidity/codegen/ExpressionCompiler.cpp:2231:	ArrayType const *arrayType = dynamic_cast<ArrayType const*>(&baseType);
libyul/backends/evm/StackLayoutGenerator.cpp:350:			CFG::BasicBlock const *block = *toVisit.begin();
```

If so, can I change files above and change right aligned '*'s to left aligned in the same PR?

aarlt commented:
Hey @hiroshitash! 

Thank you very much for your contribution!

What is the MacOS version that youre using? I was just checking the current behavoir on my machine (running MacOS Ventura 13.4.1) and it seem to ship a GNU compatible `grep` (`grep --version` returns `grep (BSD grep, GNU compatible) 2.6.0-FreeBSD.`). At least what I can tell is that the current version of the script running on MacOS v13.4.1 seem to have exactly the same behaviour of your changes. Can you run `grep --version` on your machine? I think `-E` is a GNU extension to grep, where at least the version string of `grep` on MacOS 13.4.1 is explicitly stating that it is GNU compatible. Can you also verify that youre using the `grep`
shipped with MacOS?

However, your changes look good for me. But at least for me it looks like that these changes are not needed - at least from MacOS version 13.4.1 on.

hiroshitash commented:
Hello @aarlt,

I am running the same version of MacOS (Ventura 13.4.1) with `grep (BSD grep, GNU compatible) 2.6.0-FreeBSD`.

```
# default grep on MacOS
 % grep --version
grep (BSD grep, GNU compatible) 2.6.0-FreeBSD

# installed GNU grep
 % ggrep --version
ggrep (GNU grep) 3.11
Packaged by Homebrew
```

Although BSD grep says it's compatible with GNU grep, it gives an error with a command below (while GNU grep doesn't)
```
# BSD grep shows an error
 % grep -E -v -e "return [*]" -e "^* [*]" -e "^*//.*"
grep: repetition-operator operand invalid

# GNU grep shows no errors, but some warnings
% ggrep -E -v -e "return [*]" -e "^* [*]" -e "^*//.*"
ggrep: warning: * at start of expression
ggrep: warning: * at start of expression

# GNU grep shows no warnings with modified logic
% ggrep -E -v -e "return [*]" -e ":\s*[*]" -e ".*//.*"
```

Please see the conversation [here](https://github.com/ethereum/solidity/issues/13492#issue-1363654774) for more information.

One question: With my modification, style check script still fails with legitimate style errors [here](https://github.com/ethereum/solidity/pull/14348#issuecomment-1603382593). Should I modify right-aligned '*'s in the files?

nikola-matic commented:
> I believe [these](https://app.circleci.com/pipelines/github/ethereum/solidity/30308/workflows/f88c02a0-8229-4ef1-af9d-ffaea84fc778/jobs/1345063/parallel-runs/0/steps/0-103) are valid style errors. '*'s should be left-aligned.
> 
> ```
> Coding style error:
> libsolidity/analysis/ControlFlowBuilder.cpp:587:	if (auto const *builtinFunction = m_inlineAssembly->dialect().builtin(_functionCall.functionName.name))
> libsolidity/analysis/ControlFlowBuilder.h:155:	void placeAndConnectLabel(CFGNode *_node);
> libsolidity/analysis/TypeChecker.cpp:1556:			if (auto const *variableDeclaration = dynamic_cast<VariableDeclaration const*>(identifier->annotation().referencedDeclaration))
> libsolidity/codegen/CompilerContext.h:376:	CompilerContext *m_runtimeContext;
> libsolidity/codegen/ExpressionCompiler.cpp:2231:	ArrayType const *arrayType = dynamic_cast<ArrayType const*>(&baseType);
> libyul/backends/evm/StackLayoutGenerator.cpp:350:			CFG::BasicBlock const *block = *toVisit.begin();
> ```
> 
> If so, can I change files above and change right aligned '*'s to left aligned in the same PR?

Yes please. Not really sure how these weren't caught by the script in the first place (or subsequently). Also, please rebase - I made slight changes to the check style script to have it fail when `using namespace std;` is used in specified directories.

hiroshitashir committed:
Stylecheck script fails on MacOS #13492

- Install GNU grep on macOS with brew install grep
- Use GNU grep with -E option
- The following errors fixed.

```
$scripts/check_style.sh
ggrep: warning: stray \ before /
ggrep: warning: stray \ before /
ggrep: warning: * at start of expression
ggrep: warning: * at start of expression
```

Fixes [Stylecheck script fails on MacOS #13492](https://github.com/ethereum/solidity/issues/13492)

hiroshitash commented:
Hi @aarlt,
I rebased and squashed my old and new changes into one commit.

I also made one additional change besides right-aligned '*'s:
- `\s` was replaced with `[[:space:]]` as `\s` did not behave correctly with `git grep` on my MacOS (It looks like old versions of grep do not behave correctly with `\s` according to [here](https://stackoverflow.com/questions/4233159/grep-regex-whitespace-behavior)).

All tests passed, but please let me know if anything doesn't look right.

aarlt commented:
Hey @hiroshitashir!
I'm very sorry for the very long reviewing delay! I think that your PR is good! If you're still arround, can you probably do a rebase?
Thank you very much for your contribution!

hiroshitashir commented:
Hello @aarlt,

I just rebased my branch and all tests passing. Let me know if there is anything else I can do.

cameel commented on scripts/check_style.sh:
```diff
@@ -4,6 +4,16 @@ set -eu
 
 ERROR_LOG="$(mktemp -t check_style_XXXXXX.log)"
 
+if [ "$(uname)" == "Darwin" ]; then
+    if ! command -v ggrep &> /dev/null
+    then
+        brew install grep # install GNU grep on macOS
+    fi
+    grepCommand="ggrep"
+else
+    grepCommand="grep"
+fi
```
A style check script should absolutely not modify your system by installing new software. If it requires something and it's not available, it just fail.

We actually added a portable `grep` helper recently, so please source `common.sh` and use it: https://github.com/ethereum/solidity/blob/fbbac9c1969c054071d9d6456e5151cb4ae53c19/scripts/common.sh#L317-L326

aarlt commented:
Thanks @cameel! You're right, like that it is much cleaner! Sorry for the confusion @hiroshitashir, but the proposed changes make a lot of sense. Let me know if you have any questions.

nikola-matic commented:
I'll be closing this in lieu of our current work on introducing clang-format, which essentially do away with all of the portability issues (at least as far as the check_style script is concerned).
